File: game_html5/game.js
// GLOBALS - START
var RES_X = 800;
var RES_Y = 600;
var frame = 0;

var key_left = false;
var key_right = false;
var key_space_click = false;
var key_z_click = false;

var player;
var bullets = [];
var enemies = [];
var platforms = [];
// GLOBALS - END

// Initialisation
var frameTime = 1000/60.0;
var dTime = 0;
var lastFrameTime = 0;

var mainCanvas = document.getElementById('mainCanvas');
var ctx = mainCanvas.getContext('2d');
mainCanvas.width = RES_X;
mainCanvas.height = RES_Y;
window.addEventListener('keydown', keyboardPress, false);
window.addEventListener('keyup', keyboardRelease, false);
// Initialisation - END


var Platform = function(x1, y1, x2, y2) {
    this.x1 = x1;
    this.y1 = y1;
    this.x2 = x2;
    this.y2 = y2;
};

Platform.prototype = {
    update: function() {

    },

    draw: function() {
        drawRect(this.x1, this.y1, this.x2-this.x1, this.y2-this.y1, '#00ff00');
    },
};

var Bullet = function(x, y, vx, vy, ax, ay, radius, isPlayerBullet) {
    this.x = x;
    this.y = y;
    this.vx = vx;
    this.vy = vy;
    this.ax = ax;
    this.ay = ay;
    this.radius = radius;
    this.isPlayerBullet = isPlayerBullet;
    
    this.isActive = true;
};

Bullet.prototype = {
    update: function() {
        this.vx += this.ax;
        this.vy += this.ay;
        this.x += this.vx;
        this.y += this.vy;

        if (this.isOffScreen()) {
            this.remove();
            return;
        }
    },

    remove: function() {
        this.isActive = false;
    },

    isOffScreen: function() {
        var tolerance = 100;
        return this.x < -this.radius - tolerance ||
                this.y < -this.radius - tolerance ||
                this.x > RES_X + this.radius + tolerance || 
                this.y > RES_Y + this.radius + tolerance;
    },

    draw: function() {
        drawCircle(this.x, this.y, this.radius, '#ffff00');
    },
};

var Player = function(x, y) {
    this.x = x;
    this.y = y;

    this.vx = 0;
    this.vy = 0;
    this.isOnGround = false;
    this.isFacingRight = true;
    this.isActive = true;

    this.width = 25;
    this.height = 40;
    this.speed = 3;
    this.gravity = 0.4;
    this.jumpSpeed = 10;
    this.hitboxRadius = 5;
};

Player.prototype = {
    remove: function() {
        this.isActive = false;
    },

    checkCollision: function() {
        for (var i=0; i<bullets.length; ++i) {
            var bullet = bullets[i];
            if (!bullet.isActive || bullet.isPlayerBullet) continue;
            
            var dx = bullet.x - this.x;
            var dy = bullet.y - this.y;
            var dr = this.hitboxRadius + bullet.radius;
            if (dx*dx + dy*dy <= dr*dr) {
                bullet.remove();
                this.remove();
            }
        }
    },

    physicsUpdate: function() {
        //x += vx;
        //y += vy;
        var newX = this.x + this.vx;
        var newY = this.y + this.vy;

        this.isOnGround = false;
        for (var i=0; i<platforms.length; ++i) {
            var platform = platforms[i];
            var x1 = platform.x1 - this.width/2;
            var x2 = platform.x2 + this.width/2;
            var y1 = platform.y1 - this.height/2;
            var y2 = platform.y2 + this.height/2;

            // temp variables so that I don't have to "this." everywhere
            var x = this.x;
            var y = this.y;
            var vx = this.vx;
            var vy = this.vy;

            // Collide with left wall, moving right.
            if (x <= x1 && x1 < x+vx) {
                if (vx*(y1-y) < (x1-x)*vy && (x1-x)*vy < vx*(y2-y)) {
                    newX = Math.min(x1, newX);
                    this.vx = 0;
                }
            }
            // Collide with right wall, moving left.
            if (x+vx < x2 && x2 <= x) {
                if (vx*(y2-y) < (x2-x)*vy && (x2-x)*vy < vx*(y1-y)) {
                    newX = Math.max(x2, newX);
                    this.vx = 0;
                }
            }

            // Collide with top wall, moving down. (Landing)
            if (y <= y1 && y1 < y+vy) {
                if (vy*(x1-x) <= (y1-y)*vx && (y1-y)*vx <= vy*(x2-x)) {
                    newY = Math.min(y1, newY);
                    this.vy = 0;

                    // Land
                    this.isOnGround = true;
                }
            }
            // Collide with bottom wall, moving up.
            if (y+vy < y2 && y2 <= y) {
                if (vy*(x2-x) <= (y2-y)*vx && (y2-y)*vx <= vy*(x1-x)) {
                    newY = Math.max(y2, newY);
                    this.vy = 0;
                }
            }
        }

        if (newX > RES_X - this.width/2) newX = RES_X - this.width/2;
        if (newX < this.width/2) newX = this.width/2;

        this.x = newX;
        this.y = newY;
    },

    update: function() {
        this.vx = 0;
        this.vy += this.gravity;
        if (key_left) this.vx -= this.speed;
        if (key_right) this.vx += this.speed;
        if (key_space_click) {
            if (this.isOnGround) this.vy = -this.jumpSpeed;
        }
        
        if (key_left && !key_right) this.isFacingRight = false;
        if (key_right && !key_left) this.isFacingRight = true;
        if (key_z_click) this.shootBullet();

        this.physicsUpdate();
        this.checkCollision();
    },

    shootBullet: function() {
        var bulletVx = 10;
        if (!this.isFacingRight) bulletVx *= -1;

        bullets.push(new Bullet(this.x, this.y, bulletVx, 0, 0, 0, 5, true));
    },

    draw: function() {
        drawRect(this.x - this.width/2, this.y - this.height/2, this.width, this.height, '#0000ff');
    },
};


var Enemy = function(x, y) {
    this.x = x;
    this.y = y;

    this.vx = 0;
    this.vy = 0;
    this.isActive = true;
    this.nextMovementCountdown = 0;
    this.hp = 100;
    this.shootCooldown = 30;

    this.width = 40;
    this.height = 40;
    this.speed = 2;
};

Enemy.prototype = {
    shootBullet: function() {
        var bulletVx = Math.random()*5.0 - 2.5;
        bullets.push(new Bullet(this.x, this.y, bulletVx, -5, 0, 0.1, 7, false));
    },

    takeDamage: function(damage) {
        this.hp -= damage;
        if (this.hp <= 0) this.remove();
    },

    remove: function() {
        this.isActive = false;
    },

    checkCollision: function() {
        for (var i=0; i<bullets.length; ++i) {
            var bullet = bullets[i];
            if (!bullet.isActive || !bullet.isPlayerBullet) continue;
            
            if (this.x - this.width/2 <= bullet.x + bullet.radius &&
                bullet.x - bullet.radius <= this.x + this.width/2 &&
                this.y - this.height/2 <= bullet.y + bullet.radius &&
                bullet.y - bullet.radius <= this.y + this.height/2) {

                bullet.remove();
                this.takeDamage(20);
            }
        }
    },

    updateVelocity: function() {
        if (this.nextMovementCountdown > 0) {
            --this.nextMovementCountdown;
            return;
        } else {
            this.nextMovementCountdown = Math.random()*120 + 60;
        }

        var angle = Math.random() * (2*3.1415926535);
        this.vx = this.speed*Math.cos(angle);
        this.vy = this.speed*Math.sin(angle);

        if (this.x > RES_X && this.vx > 0) this.vx *= -1;
        if (this.x < 0 && this.vx < 0) this.vx *= -1;
        if (this.y > 2*RES_Y/3 && this.vy > 0) this.vy *= -1;
        if (this.y < RES_Y/4 && this.vy < 0) this.vy *= -1;
    },

    updateShooting: function() {
        if (this.shootCooldown <= 0) {
            this.shootBullet();
            this.shootCooldown = 20;
        } else {
            --this.shootCooldown;
        }
    },

    update: function() {
        this.updateVelocity();
        this.updateShooting();
        this.x += this.vx;
        this.y += this.vy;
        this.checkCollision();
    },

    draw: function() {
        drawRect(this.x - this.width/2, this.y - this.height/2, this.width, this.height, '#ff0000');
    },
};



function initialiseGame() {
    bullets = [];
    enemies = [];
    platforms = [];

    player = new Player(RES_X/4, RES_Y/4);
    platforms.push(new Platform(120,310,220,330));
    platforms.push(new Platform(000,420,380,440));
    platforms.push(new Platform(430,360,550,380));
    platforms.push(new Platform(600,420,800,440));

    frame = 0;
}

function updateGame() {
    if (player.isActive) player.update();

    for (var i=0; i<bullets.length; ++i) {
        if (bullets[i].isActive) bullets[i].update();
    }

    for (var i=0; i<enemies.length; ++i) {
        if (enemies[i].isActive) enemies[i].update();
    }

    for (var i=0; i<platforms.length; ++i) {
        platforms[i].update();
    }

    if (frame%300 == 0) {
        enemies.push(new Enemy(Math.random()*RES_X, Math.random()*RES_Y/3));
    }

    resetKeyClicks();
    ++frame;
}

function drawGameFrame() {
    if (player.isActive) player.draw();

    for (var i=0; i<bullets.length; ++i) {
        if (bullets[i].isActive) bullets[i].draw();
    }

    for (var i=0; i<enemies.length; ++i) {
        if (enemies[i].isActive) enemies[i].draw();
    }

    for (var i=0; i<platforms.length; ++i) {
        platforms[i].draw();
    }
}

function keyboardPress(e) {
    console.log('Pressed ' + e.keyCode);
    switch(e.keyCode) {
        case 37: key_left = true; break;
        case 39: key_right = true; break;
        case 32: key_space_click = true; break;
        case 90: key_z_click = true; break;
        case 82: initialiseGame(); break;
        default: return;
    }
    e.preventDefault();
}

function keyboardRelease(e) {
    switch(e.keyCode) {
        case 37: key_left = false; break;
        case 39: key_right = false; break;
        default: return;
    }
    e.preventDefault();
}

function resetKeyClicks() {
    key_space_click = false;
    key_z_click = false;
}

// This is called in an infinite while loop.
function gameLoop(time){
    if (time !== undefined) {
        dTime += (time - lastFrameTime);
        lastFrameTime = time;
        
        // Safeguard (slowdown) to prevent game from lagging to death
        if (dTime > 5*frameTime) dTime = 5*frameTime;

        // Update Frame
        while (dTime > frameTime) {
            updateGame();
            dTime -= frameTime;
        }
        
        // Draw Frame
        clearScreen();
        drawGameFrame();
    }    
    window.requestAnimationFrame(gameLoop);
}

function main() {
    initialiseGame();
    gameLoop();
}

File: game_pygame/game.py
import sys, pygame
import random
import math
FRAME_UPDATE_EVENT = pygame.USEREVENT+1

# GLOBALS - START
screen = None
RES_X = 800
RES_Y = 600
frame = 0

key_left = False
key_right = False
key_space_click = False
key_z_click = False

player = None
bullets = []
enemies = []
platforms = []
# GLOBALS - END

class Platform(object):
    def __init__(self, x1, y1, x2, y2):
        self.x1 = x1
        self.y1 = y1
        self.x2 = x2
        self.y2 = y2

    def update(self):
        pass

    def draw(self):
        pygame.draw.rect(screen, (0,255,0),
            (self.x1, self.y1, self.x2-self.x1, self.y2-self.y1))

class Bullet(object):
    def __init__(self, x, y, vx, vy, ax, ay, radius, isPlayerBullet):
        self.x = x
        self.y = y
        self.vx = vx
        self.vy = vy
        self.ax = ax
        self.ay = ay
        self.radius = radius
        self.isPlayerBullet = isPlayerBullet
        
        self.isActive = True

    def update(self):
        self.vx += self.ax
        self.vy += self.ay
        self.x += self.vx
        self.y += self.vy

        if self.isOffScreen():
            self.remove()
            return

    def remove(self):
        self.isActive = False

    def isOffScreen(self):
        tolerance = 100
        return self.x < -self.radius - tolerance or \
                self.y < -self.radius - tolerance or \
                self.x > RES_X + self.radius + tolerance or  \
                self.y > RES_Y + self.radius + tolerance

    def draw(self):
        pygame.draw.circle(screen, (255,255,0), (int(self.x), int(self.y)), self.radius)

class Player(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y

        self.vx = 0
        self.vy = 0
        self.isOnGround = False
        self.isFacingRight = True
        self.isActive = True

        self.width = 25
        self.height = 40
        self.speed = 3
        self.gravity = 0.4
        self.jumpSpeed = 10
        self.hitboxRadius = 5

    def remove(self):
        self.isActive = False

    def checkCollision(self):
        for bullet in bullets:
            if not bullet.isActive or bullet.isPlayerBullet: continue
            
            dx = bullet.x - self.x
            dy = bullet.y - self.y
            dr = self.hitboxRadius + bullet.radius
            if dx*dx + dy*dy <= dr*dr:
                bullet.remove()
                self.remove()

    def physicsUpdate(self):
        # x += vx
        # y += vy
        newX = self.x + self.vx
        newY = self.y + self.vy

        self.isOnGround = False
        for platform in platforms:
            x1 = platform.x1 - self.width/2
            x2 = platform.x2 + self.width/2
            y1 = platform.y1 - self.height/2
            y2 = platform.y2 + self.height/2

            # temp variables so that I don't have to "self." everywhere.
            x = self.x
            y = self.y
            vx = self.vx
            vy = self.vy

            # Collide with left wall, moving right.
            if x <= x1 and x1 < x+vx:
                if vx*(y1-y) < (x1-x)*vy and (x1-x)*vy < vx*(y2-y):
                    newX = min(x1, newX)
                    self.vx = 0
            # Collide with right wall, moving left.
            if x+vx < x2 and x2 <= x:
                if vx*(y2-y) < (x2-x)*vy and (x2-x)*vy < vx*(y1-y):
                    newX = max(x2, newX)
                    self.vx = 0

            # Collide with top wall, moving down. (Landing)
            if y <= y1 and y1 < y+vy:
                if vy*(x1-x) <= (y1-y)*vx and (y1-y)*vx <= vy*(x2-x):
                    newY = min(y1, newY)
                    self.vy = 0

                    # Land
                    self.isOnGround = True
            # Collide with bottom wall, moving up.
            if y+vy < y2 and y2 <= y:
                if vy*(x2-x) <= (y2-y)*vx and (y2-y)*vx <= vy*(x1-x):
                    newY = max(y2, newY)
                    self.vy = 0

        if newX > RES_X - self.width/2: newX = RES_X - self.width/2
        if newX < self.width/2: newX = self.width/2

        self.x = newX
        self.y = newY

    def update(self):
        self.vx = 0
        self.vy += self.gravity
        if key_left: self.vx -= self.speed
        if key_right: self.vx += self.speed
        if key_space_click:
            if self.isOnGround: self.vy = -self.jumpSpeed
        
        if key_left and not key_right: self.isFacingRight = False
        if key_right and not key_left: self.isFacingRight = True
        if key_z_click: self.shootBullet()

        self.physicsUpdate()
        self.checkCollision()

    def shootBullet(self):
        bulletVx = 10
        if not self.isFacingRight: bulletVx *= -1

        bullets.append(Bullet(self.x, self.y, bulletVx, 0, 0, 0, 5, True))

    def draw(self):
        pygame.draw.rect(screen, (0,0,255),
            (self.x - self.width/2, self.y - self.height/2, self.width, self.height))


class Enemy(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y

        self.vx = 0
        self.vy = 0
        self.isActive = True
        self.nextMovementCountdown = 0
        self.hp = 100
        self.shootCooldown = 30

        self.width = 40
        self.height = 40
        self.speed = 2

    def shootBullet(self):
        bulletVx = random.random()*5.0 - 2.5
        bullets.append(Bullet(self.x, self.y, bulletVx, -5, 0, 0.1, 7, False))

    def takeDamage(self, damage):
        self.hp -= damage
        if self.hp <= 0: self.remove()

    def remove(self):
        self.isActive = False

    def checkCollision(self):
        for bullet in bullets:
            if not bullet.isActive or not bullet.isPlayerBullet: continue
            
            if self.x - self.width/2 <= bullet.x + bullet.radius and \
                bullet.x - bullet.radius <= self.x + self.width/2 and \
                self.y - self.height/2 <= bullet.y + bullet.radius and \
                bullet.y - bullet.radius <= self.y + self.height/2:

                bullet.remove()
                self.takeDamage(20)

    def updateVelocity(self):
        if self.nextMovementCountdown > 0:
            self.nextMovementCountdown -= 1
            return
        else:
            self.nextMovementCountdown = random.randrange(60,180)

        angle = random.random() * (2*math.pi)
        self.vx = self.speed*math.cos(angle)
        self.vy = self.speed*math.sin(angle)

        if self.x > RES_X and self.vx > 0: self.vx *= -1
        if self.x < 0 and self.vx < 0: self.vx *= -1
        if self.y > 2*RES_Y/3 and self.vy > 0: self.vy *= -1
        if self.y < RES_Y/4 and self.vy < 0: self.vy *= -1

    def updateShooting(self):
        if self.shootCooldown <= 0:
            self.shootBullet()
            self.shootCooldown = 20
        else:
            self.shootCooldown -= 1

    def update(self):
        self.updateVelocity()
        self.updateShooting()
        self.x += self.vx
        self.y += self.vy
        self.checkCollision()

    def draw(self):
        pygame.draw.rect(screen, (255,0,0),
            (self.x - self.width/2, self.y - self.height/2, self.width, self.height))


def initialiseGame():
    global bullets, enemies, platforms, player, frame
    bullets = []
    enemies = []
    platforms = []

    player = Player(RES_X/4, RES_Y/4)
    platforms.append(Platform(120,310,220,330))
    platforms.append(Platform(000,420,380,440))
    platforms.append(Platform(430,360,550,380))
    platforms.append(Platform(600,420,800,440))

    frame = 0

def updateGame():
    global frame

    if player.isActive: player.update()

    for bullet in bullets:
        if bullet.isActive: bullet.update()

    for enemy in enemies:
        if enemy.isActive: enemy.update()

    for platform in platforms:
        platform.update()

    if frame%300 == 0:
        enemies.append(Enemy(random.randrange(0,RES_X), random.randrange(0,RES_Y//3)))

    resetKeyClicks()
    frame += 1

def drawGameFrame():
    if player.isActive: player.draw()

    for bullet in bullets:
        if bullet.isActive: bullet.draw()

    for enemy in enemies:
        if enemy.isActive: enemy.draw()

    for platform in platforms:
        platform.draw()

def keyDown(key):
    global key_left, key_right, key_space_click, key_z_click
    if key == pygame.K_LEFT: key_left = True
    elif key == pygame.K_RIGHT: key_right = True
    elif key == pygame.K_SPACE: key_space_click = True
    elif key == pygame.K_z: key_z_click = True
    elif key == pygame.K_r: initialiseGame()

def keyUp(key):
    global key_left, key_right
    if key == pygame.K_LEFT: key_left = False
    elif key == pygame.K_RIGHT: key_right = False

def resetKeyClicks():
    global key_space_click, key_z_click
    key_space_click = False
    key_z_click = False

def processEvent(event):
    if event.type == FRAME_UPDATE_EVENT:
        updateGame()
    if event.type == pygame.QUIT:
        sys.exit()
    elif event.type == pygame.KEYDOWN:
        keyDown(event.key)
    elif event.type == pygame.KEYUP:
        keyUp(event.key)

def main():
    pygame.init()
    global screen
    screen = pygame.display.set_mode((RES_X, RES_Y))
    pygame.display.set_caption('This is not a game. This is serious business.')
    
    frameTime = 1000/60 # in ms
    dTime = 0.0
    lastFrameTicks = pygame.time.get_ticks()
    
    initialiseGame()
    while True:
        currentFrameTicks = pygame.time.get_ticks()
        dTime += (currentFrameTicks - lastFrameTicks)
        lastFrameTicks = currentFrameTicks
        
        for event in pygame.event.get():
            processEvent(event)
        
        # Safeguard (slowdown) to prevent game from lagging to death
        if dTime > 5*frameTime: dTime = 5*frameTime
        
        # Update Frame
        while dTime > frameTime:
            dTime -= frameTime
            updateGame()
        
        # Draw Frame
        screen.fill((0,0,0))
        drawGameFrame()
        pygame.display.flip()

if __name__ == '__main__':
    main()